docs/proposals/version_documentation.md

```diff
@@ -0,0 +1,145 @@
+2. A solid and robust Documentation structure.
+3. Design a workflow for managing docs that integrates with Thanos Git workflow ie updating corresponding docs on pull requests, cherry-picks, etc.
+
+# Motivation
+
+Many Users (mostly developers) would often want to look through the docs of previous releases, but searching for it manually by looking through the entire GitHub repository is time-consuming and ineffective. Because the site is built by rendering the markdown files from the tip of master of the Thanos repository, it causes confusion for users, as breaking changes are often merged into master. The verioning plugin allows the user to access the docs of both latest and previous releases on the Thanos documentation page, and it also allows the developer to fetch, write, and fix the docs of both latest and previous releases.
+
+
+# Requirements
+
+### User Story (Latest)
+
+* As a Thanos developer, I want to be able to just write docs for my current `master` version
+* As a Thanos developer, I want to build website within versioned docs with a single action
+* As a Thanos developer, I don't want to store in a single repository version duplicated docs
+
+### User Story (Previous release)
+
+* As a Thanos developer, I want to be able to fix docs for certain previous release.
+* As a Thanos user, I want to be able to read docs for older version of Thanos
+
+### Use Case
+
+#### Users
+Thanos Developers, Hugo developers, and General Users
+
+#### Precondition
+The user visits the Thanos Documentation Page
+
+#### Basic Course Of Events
+
+1. The User indicates that the site is to show a list of docs (both latest and previous releases) by clicking the version picker (dropdown)
+
+2. The site responds by displaying a list of versions allowing the user to make a selection.
+
+3. User makes selection and the required docs renders on the page/site.
+
+
+#### Postcondition
+
+The site now has a version picker containing a list of versioned docs consisting of both latest and older versions.
+
+
+# Proposed Solution
+
+## 1. Version Picker
+
+Currently, the documentation is residing under the docs/ folder of Thanos repository. It is built by Hugo. It will have a proper drop-down menu just like prometheus [`drop-down menu`](https://prometheus.io/docs/introduction/overview/) which will enable proper versioning. This user facing tool will be built using HTML and CSS
+
+
+## 2. Documentation Structure
+
+We want to propose a method called "Directory Sub Branching".
+Directory Sub branching means creating different sub branches in the content folder of the Thanos repository. Since the current architecture of Hugo is this:
+```
+|- archetypes
+|- config.toml
+|- content
+|- data
+|- layouts
+|- static
+|- themes
+
+```
+
+You can add aditional versioned content in different folders within the content directory. For example:
+```
+
+content
```

Comment:
I think I know what you mean but from this perspective it looks like it's a manual step (because you talk about automation in next section). Wonder if we can make it clear that docs are updates as usual now in `docs` on master, then releases are done via tags etc (:

Reply:
It's not manual work anymore, so it's amazing but still we just store duplicated a lot (because we have already this content in git).

What about:

1. Just do everything on `/docs` on master
1. Release 0.14
1. Run some `make web / web-serve`
   1. Before anything, we generate those docs 
   1. ... rest of web command is done
  
NOTE: generated docs are not committed, just temporarily built, same as `public` dir.

Reply:
## Open question: 

* Specify what docs?
   * .. incompatible way?
   * Flexible configuration file:
          * docs.yaml 

### Option 1:

```
versioned:
   - docs/components/*.md
   - docs/storage2221241.md
```

   * We could have this file on master, so the current `make web` will 
          1. Parse on master docs.yaml and decide.
          1. Go to `release-0.12`
               1. Is there docs.yaml? Yes - parse and decide
                   No - use docs.yaml from master   

### Option 2 (chosen):

Single file
```
versioned:
   default:
   - docs/components/*.md
   - docs/storage.md
   
  overrides:
     release-0.12:
     - docs/storage2221241.md
```
~Problem: Manual edit of each release.~

### Option 3:

Change directory structure back in time (: 

* How we do fixes to release docs for older release e.g v0.12? (: 

What about editing `release-0.12` and commit. Then the tool fetches the latest on this branch and use it to generate docs.

"release is not just a certain tag with sem ver - it's actually a special git branch. For Thanos `release-.*`

We expect docs for minor version is what matters - not breaking with patch releases /  rc.